docs/proposal/20200506-single-controller-multitenancy.md

```diff
@@ -0,0 +1,869 @@
+assumption using another method such as Kiam.
+
+#### Caching and handling refresh of credentials
+
+For handling many accounts, the number of calls to the STS service must be minimised. This is currently implemented as a
+cache on cluster key in Cluster API Provider AWS.
+
+#### AWS Cloud Provider behaviour with regards to cluster names
+
+At present, multiple clusters deployed into the same AWS account with the same
+cluster name cause the AWS Cloud Provider to malfunction due to the way
+tags are looked up. This has driven some concern to allowing multiple
+namespaces access to the same AWS account.
+
+However, since it is possible to configure multiple credentials against the same
+account, there's no easy way to guarantee uniqueness.
+
+This could be mitigated by having CAPA call `STS::GetCallerIdentity` on entry
+to the cluster, and then comparing against all current clusters declared across
+the management cluster. We would probably need to store the AccountID into
+the AWSCluster object and index the field.
+
+A better approach maybe to modify the cloud provider to support customisation
+of tag lookups such that multiple clusters with the same name, but in
+different namespaces do not break the AWS Cloud Provider.
+
+## Upgrade Strategy
+
+The data changes are additive and optional, so existing AWSCluster specifications will continue to reconcile as before.
+These changes will only come into play when an IAM role is provided in the new field in AWSClusterSpec. Upgrades to
+versions with this new field will be broken
+
+## Additional Details
+
+### Test Plan
+
+* Unit tests to validate that the cluster controller can reconcile an AWSCluster when IAMRoleARN field is nil, or provided.
+* Unit tests to ensure pre-flight checks are performed relating to IAM role assumption when IAMRoleARN is provided.
+* Propose performing an initial sts:AssumeRole call and fail pre-flight if this fails.
+* e2e test for role assumption ([NFR12](#NFR12)).
+* If it can be supported in Prow environment, additional e2e test for OIDC-based role assumption.
+
+### Graduation Criteria
+
+#### Alpha
+
+* Support role assumption with external ID as a aws-sdk-go `credentials.Provider`.
+
+#### Beta
+
+* Support OIDC and chained role assumption.
+* Admission controller validation for secrets of type.
+* Full e2e coverage.
+
+#### Stable
+
+* Two releases since beta.
+
+These may be defined in terms of API maturity, or as something else. Initial proposal should keep this high-level with a
+focus on what signals will be looked at to determine graduation.
+
+### Version Skew Strategy
+
+Most of the version skew is contained within the secret type. We explicitly add a `v1alpha3` to the secret type allowing
+the format to change over time. However, we do not have the benefit of hub and spoke storage conversion that would exist
+with a CRD. We propose, that in the event of a secret type signature change:
```

Comment:
Is there anyway for us to leverage the client-go config format parsing here to avoid having to embed a typed config that we own here?

Reply:
Another case to stick these in their own API group. If there's a change we can increment the version and deal with them with their own conversion webhooks.

Reply:
But otherwise, config parsing is stuck in an internal package in the aws sdk